miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/storage/MiniAppStorage.kt

```diff
@@ -43,6 +49,33 @@ internal class MiniAppStorage(
     @VisibleForTesting
     fun getFileName(file: String) = urlToFileInfoParser.getFileName(file)
 
-    fun getSavePathForApp(appId: String, versionId: String) =
-        "${basePath.path}/$SUB_DIR_MINIAPP/$appId/$versionId"
+    @VisibleForTesting
+    internal fun getBasePathHostApp() = basePath.path
+
+    @VisibleForTesting
+    internal fun getRootPathMiniApp() = "${getBasePathHostApp()}/$SUB_DIR_MINIAPP/"
+
+    @VisibleForTesting
+    internal fun getPathMiniApp(appId: String) = "${getRootPathMiniApp()}/$appId/"
+
+    fun getPathMiniAppVersion(appId: String, versionId: String) = "${getPathMiniApp(appId)}$versionId"
+
+    suspend fun removeOutdatedVersionApp(
+        appId: String,
+        latestVersionId: String,
+        appPath: String = getPathMiniApp(appId)
+    ) =
+            withContext(Dispatchers.IO) {
+                launch(Job()) {
+                    val parentFile = File(appPath)
+                    if (parentFile.isDirectory && parentFile.listFiles() != null) {
+                        flow {
+                            parentFile.listFiles()?.forEach { file ->
+                                if (!file.absolutePath.endsWith(latestVersionId))
+                                    emit(file)
+                            }
+                        }.collect { file -> file.deleteRecursively() }
```

Comment:
Is it possible for the delete to fail? Should we add some error handling here?

On iOS, it just logs a debug message when deletion fails.

Reply:
There is an exception `SecurityException` which denies the access of file. I think I'll implement the general exception handling just like iOS.